wcfsetup/install/files/lib/http/middleware/TriggerBackgroundQueue.class.php

```diff
@@ -0,0 +1,58 @@
+<?php
+
+namespace wcf\http\middleware;
+
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+use wcf\http\LegacyPlaceholderResponse;
+use wcf\system\background\BackgroundQueueHandler;
+use wcf\system\WCFACP;
+
+/**
+ * Adds 'woltlab-background-queue-check: yes' to the response
+ * whenever a check was requested.
+ *
+ * @author Alexander Ebert
+ * @copyright 2001-2022 WoltLab GmbH
+ * @license GNU Lesser General Public License <http://opensource.org/licenses/lgpl-license.php>
+ * @package WoltLabSuite\Core\Http\Middleware
+ * @since 6.0
+ */
+final class TriggerBackgroundQueue implements MiddlewareInterface
+{
+    private const HEADER_NAME = 'woltlab-background-queue-check';
+    private const HEADER_VALUE = 'yes';
+
+    private readonly BackgroundQueueHandler $backgroundQueueHandler;
+
+    public function __construct()
+    {
+        $this->backgroundQueueHandler = BackgroundQueueHandler::getInstance();
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
+    {
+        $response = $handler->handle($request);
+        if (
+            \class_exists(WCFACP::class, false)
+            || !$this->backgroundQueueHandler->hasPendingCheck()
+        ) {
+            return $response;
+        }
+
+        if ($response instanceof LegacyPlaceholderResponse) {
+            \header(
+                \sprintf('%s: %s', self::HEADER_NAME, self::HEADER_VALUE)
+            );
```

Comment:
This might be unsafe for two reasons:

- The non-PSR-15 controller might’ve already emitted some output and thus headers are already sent, causing an error.
- The non-PSR-15 controller might call `exit;`, preventing any middlewares from running in the response path (this is what happens in AJAXInvokeAction and thus AJAXProxyAction).

For `LegacyPlaceholderResponse` the rule is: Don’t touch, here be dragons.

-----------

Instead for compatibility the following might be required:

- `BackgroundQueueHandler::forceCheck()` adds the header on a best-effort basis (with `@\header()`, including the `@`).
- This middleware removes the header again with `\header_remove()` and then adds it to the PSR-7 response object to be emitted by the PSR-7 pipeline, similarly to `PreventMimeSniffing`.

-------------------------

Alternatively the legacy AJAXProxyAction continues to add the metadata to the JSON itself and the header is only used for new-style controllers. Technically `AJAXInvokeAction` could also be adjusted to emit a PSR-7 response in `AJAXInvokeAction::sendResponse()` but that might be considered a break.